File: src/main/java/de/spries/fleetcommander/model/facade/PlayerSpecificPlanet.java
package de.spries.fleetcommander.model.facade;

import java.util.List;
import java.util.stream.Collectors;

import org.apache.commons.lang3.builder.EqualsBuilder;
import org.apache.commons.lang3.builder.HashCodeBuilder;

import de.spries.fleetcommander.model.core.Player;
import de.spries.fleetcommander.model.core.universe.FactorySite;
import de.spries.fleetcommander.model.core.universe.HasCoordinates;
import de.spries.fleetcommander.model.core.universe.Planet;

public class PlayerSpecificPlanet implements HasCoordinates {

	private Planet originalPlanet;
	private Player viewingPlayer;

	public PlayerSpecificPlanet(Planet originalPlanet, Player viewingPlayer) {
		this.originalPlanet = originalPlanet;
		this.viewingPlayer = viewingPlayer;
	}

	public int getId() {
		return originalPlanet.getId();
	}

	@Override
	public int getX() {
		return originalPlanet.getX();
	}

	@Override
	public int getY() {
		return originalPlanet.getY();
	}

	public String getPlanetClass() {
		return isInhabitedByMe() ? originalPlanet.getPlanetClass().name() : "?";
	}

	public boolean isMyHomePlanet() {
		return originalPlanet.isHomePlanetOf(viewingPlayer);
	}

	public boolean isInhabitedByMe() {
		return originalPlanet.isInhabitedBy(viewingPlayer);
	}

	public boolean isKnownAsEnemyPlanet() {
		return originalPlanet.isKnownAsEnemyPlanet(viewingPlayer);
	}

	public boolean isUnderAttack() {
		if (isInhabitedByMe()) {
			return originalPlanet.isUnderAttack();
		}
		return false;
	}

	public boolean isJustInhabited() {
		if (isInhabitedByMe()) {
			return originalPlanet.isJustInhabited();
		}
		return false;
	}

	public boolean canBuildFactory() {
		return originalPlanet.canBuildFactory(viewingPlayer);
	}

	public int getShipCount() {
		if (isInhabitedByMe()) {
			return originalPlanet.getShipCount();
		}
		return 0;
	}

	public int getIncomingShipCount() {
		return originalPlanet.getIncomingShipCount(viewingPlayer);
	}

	public FactorySite getFactorySite() {
		if (isInhabitedByMe()) {
			return originalPlanet.getFactorySite();
		}
		return null;
	}

	public void buildFactory() {
		originalPlanet.buildFactory(viewingPlayer);
	}

	public void changeProductionFocus(int focus) {
		originalPlanet.setProductionFocus(focus, viewingPlayer);
	}

	@Override
	public boolean equals(Object obj) {
		return EqualsBuilder.reflectionEquals(this, obj);
	}

	@Override
	public int hashCode() {
		return HashCodeBuilder.reflectionHashCode(this);
	}

	public static List<PlayerSpecificPlanet> filterMyPlanets(List<PlayerSpecificPlanet> allPlanets) {
		return allPlanets.stream().filter(p -> p.isInhabitedByMe())
				.collect(Collectors.toList());
	}

	protected static PlayerSpecificPlanet convert(Planet planet, Player viewingPlayer) {
		return new PlayerSpecificPlanet(planet, viewingPlayer);
	}

	protected static List<PlayerSpecificPlanet> convert(List<Planet> planets, Player viewingPlayer) {
		return planets.stream().map(p -> convert(p, viewingPlayer))
				.collect(Collectors.toList());
	}
}

File: src/main/kotlin/de/spries/fleetcommander/model/facade/PlayerSpecificPlanet.kt
package de.spries.fleetcommander.model.facade

import de.spries.fleetcommander.model.core.Player
import de.spries.fleetcommander.model.core.universe.HasCoordinates
import de.spries.fleetcommander.model.core.universe.Planet

open class PlayerSpecificPlanet(private val originalPlanet: Planet, private val viewingPlayer: Player)
    : HasCoordinates(originalPlanet.x, originalPlanet.y) {

    fun getId() = originalPlanet.id

    fun getPlanetClass() = if (isInhabitedByMe()) originalPlanet.getPlanetClass().name else "?"

    fun isMyHomePlanet() = originalPlanet.isHomePlanetOf(viewingPlayer)

    fun isHomePlanet() = originalPlanet.isHomePlanetOf(viewingPlayer) || originalPlanet.isHomePlanet() && isKnownAsEnemyPlanet()

    fun isInhabitedByMe() = originalPlanet.isInhabitedBy(viewingPlayer)

    fun isKnownAsEnemyPlanet() = if (TestMode.TEST_MODE && viewingPlayer.isHumanPlayer()) {
        originalPlanet.isInhabited() && !originalPlanet.isInhabitedBy(viewingPlayer)
    } else originalPlanet.isKnownAsEnemyPlanet(viewingPlayer)

    fun isUnderAttack() = if (isInhabitedByMe()) {
        originalPlanet.isUnderAttack()
    } else false

    fun isJustInhabited() = if (isInhabitedByMe()) {
        originalPlanet.isJustInhabited()
    } else false

    fun getShipCount() = if (isInhabitedByMe() || TestMode.TEST_MODE && viewingPlayer.isHumanPlayer()) {
        originalPlanet.getShipCount()
    } else 0

    fun getIncomingShipCount() = originalPlanet.getIncomingShipCount(viewingPlayer)

    fun getFactorySite() = if (isInhabitedByMe() || TestMode.TEST_MODE && viewingPlayer.isHumanPlayer()) {
        originalPlanet.getFactorySite()
    } else null

    fun canBuildFactory(): Boolean {
        return originalPlanet.canBuildFactory(viewingPlayer)
    }

    fun buildFactory() {
        originalPlanet.buildFactory(viewingPlayer)
    }

    fun changeProductionFocus(focus: Int) {
        originalPlanet.setProductionFocus(focus, viewingPlayer)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as PlayerSpecificPlanet

        if (originalPlanet != other.originalPlanet) return false
        if (viewingPlayer != other.viewingPlayer) return false

        return true
    }

    override fun hashCode(): Int {
        var result = originalPlanet.hashCode()
        result = 31 * result + viewingPlayer.hashCode()
        return result
    }


}
